File: backend/src/main/java/org/icpclive/events/WF/WFTeamInfo.java
package org.icpclive.events.WF;

import org.icpclive.events.RunInfo;
import org.icpclive.events.TeamInfo;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;

/**
 * Created by Meepo on 3/5/2018.
 */
public class WFTeamInfo implements TeamInfo {

    protected ArrayList<ArrayList<RunInfo>> problem_runs;

    public int id = -1;
    public int rank;
    public String name;

    public int solved;
    public int penalty;
    public long lastAccepted;
    public HashSet<String> groups;

    public String shortName;

    public String hashTag;

    public WFTeamInfo(int problems) {
        problem_runs = new ArrayList<>(problems);
        for (int i = 0; i < problems; i++) {
            problem_runs.add(new ArrayList<>());
        }
        groups = new HashSet<>();
    }

    public WFTeamInfo(WFTeamInfo teamInfo) {
        this(teamInfo.getRuns().size());
        id = teamInfo.id;
        rank = teamInfo.rank;
        name = teamInfo.name;

        groups = new HashSet<>(teamInfo.groups);
        shortName = teamInfo.shortName;
    }

    public WFTeamInfo copy() {
        WFTeamInfo teamInfo = new WFTeamInfo(problem_runs.size());
        teamInfo.id = id;
        teamInfo.rank = rank;
        teamInfo.name = name;
        teamInfo.groups = new HashSet<>(groups);
        teamInfo.shortName = shortName;

        return teamInfo;
    }

    @Override
    public int getId() {
        return id;
    }

    @Override
    public int getRank() {
        return rank;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getShortName() {
        return shortName;
    }

    @Override
    public String getAlias() {
        return (id + 1) + "";
    }

    @Override
    public HashSet<String> getGroups() {
        return groups;
    }

    @Override
    public int getPenalty() {
        return penalty;
    }

    @Override
    public int getSolvedProblemsNumber() {
        return solved;
    }

    public List<? extends List<? extends RunInfo>> getRuns() {
        return problem_runs;
    }

    public long getLastAccepted() {
        return lastAccepted;
    }

    public String getHashTag() {
        return hashTag;
    }

    public void addRun(RunInfo run, int problemId) {
        ArrayList<RunInfo> runs = problem_runs.get(problemId);
        synchronized (runs) {
            runs.add(run);
        }
    }

    public String toString() {
        return String.format("%03d", id + 1) + ". " + shortName;
    }
}

File: backend/src/main/kotlin/org/icpclive/cds/wf/WFTeamInfo.kt
package org.icpclive.cds.wf

import org.icpclive.api.MediaType
import org.icpclive.cds.RunInfo
import org.icpclive.cds.TeamInfo

/**
 * Created by Meepo on 3/5/2018.
 */
open class WFTeamInfo(problems: Int) : TeamInfo {
    protected var problem_runs: ArrayList<ArrayList<RunInfo>>
    override var id = -1
    override lateinit var name: String
    override lateinit var groups: HashSet<String>
    override lateinit var shortName: String
    override var hashTag: String? = null
    override val medias: Map<MediaType, String> = emptyMap()

    init {
        problem_runs = ArrayList(problems)
        for (i in 0 until problems) {
            problem_runs.add(ArrayList())
        }
        groups = HashSet()
    }

    constructor(teamInfo: WFTeamInfo) : this(teamInfo.runs.size) {
        id = teamInfo.id
        name = teamInfo.name
        groups = HashSet(teamInfo.groups)
        shortName = teamInfo.shortName
    }

    override val contestSystemId: String
        get() = (id + 1).toString() + ""
    val runs: List<List<RunInfo>>
        get() = problem_runs

    fun addRun(run: RunInfo, problemId: Int) {
        val runs = problem_runs[problemId]
        synchronized(runs) { runs.add(run) }
    }

    override fun toString(): String {
        return String.format("%03d", id + 1) + ". " + shortName
    }
}